UefiDbgExt/modules.cpp

```diff
@@ -81,77 +81,108 @@ FindModuleBackwards (
   return Result;
 }
 
-HRESULT CALLBACK
+HRESULT
 loadmodules (
-  PDEBUG_CLIENT4  Client,
-  PCSTR           args
+  ULONG64  SystemTableAddr
   )
 {
-  ULONG64  HeaderAddress;
-  UINT32   TableSize;
-  ULONG64  Table;
-  ULONG64  Entry;
-  ULONG64  NormalImage;
-  ULONG64  ImageProtocol;
-  UINT64   ImageBase;
-  ULONG    Index;
-  CHAR     Command[512];
-  ULONG64  Base;
-
-  INIT_API ();
-
-  UNREFERENCED_PARAMETER (args);
+  UINT32                       TableSize;
+  ULONG64                      Table;
+  EFI_DEBUG_IMAGE_INFO         *Entry;
+  EFI_DEBUG_IMAGE_INFO_NORMAL  *NormalImage;
+  EFI_LOADED_IMAGE_PROTOCOL    *ImageProtocol;
+  UINT64                       ImageBase;
+  ULONG                        Index;
+  CHAR                         Command[512];
+  ULONG64                      Base;
+  ULONG                        BytesRead = 0;
 
   //
   // TODO: Add support for PEI & MM
   //
 
-  if (gUefiEnv != DXE) {
-    dprintf ("Only supported for DXE!\n");
-    return ERROR_NOT_SUPPORTED;
+  EFI_SYSTEM_TABLE                   SystemTable;
+  EFI_CONFIGURATION_TABLE            *ConfigTable;
+  GUID                               DebugImageInfoTableGuid    = EFI_DEBUG_IMAGE_INFO_TABLE_GUID;
+  EFI_DEBUG_IMAGE_INFO_TABLE_HEADER  *DebugImageInfoTableHeader = NULL;
+
+  // Read the EFI_SYSTEM_TABLE structure from the provided address
+  if (!ReadMemory (SystemTableAddr, &SystemTable, sizeof (SystemTable), &BytesRead) || (BytesRead != sizeof (SystemTable))) {
+    dprintf ("Failed to read EFI_SYSTEM_TABLE at %llx\n", SystemTableAddr);
+    return ERROR_NOT_FOUND;
   }
 
-  HeaderAddress = GetExpression ("&mDebugInfoTableHeader");
-  if (HeaderAddress == NULL) {
-    dprintf ("Failed to find mDebugInfoTableHeader!\n");
+  // Iterate through the configuration tables to find the debug image info table
+  ConfigTable = SystemTable.ConfigurationTable;
+  for (UINT64 i = 0; i < SystemTable.NumberOfTableEntries; i++) {

```

Comment:
Is there a reason for `i` to be a `UINT64` here instead of `SIZE_T`?

Reply:
I see `NumberOfTableEntries` is `UINT64`. So, that's fine for indexing table entries.